### DISCUSSION TITLE
Do we have api to recall the center vector against request?

### DISCUSSION THREAD

**Discussion by godlockin** [milvus-io/milvus#6673] [Q&A and General discussion] [answered]
Can I recall the kernel vector whether it was exists or not by querying with a vector against a dataset?

I setup a dataset with assigned the number of kernels, and insert several vectors.

If I were right, milvus will maintain several clusters within the dataset (the number of clusters I set during the dataset init)

The insert process is something like indexing the new vector into an exist cluster.

The recall process is like find the cluster which the vector in request is belongs to, and return the topK vectors belongs to the cluster.

My question is, how can I get the "real" center vector belongs the same cluster with the request one? And does the center kernel always be one of the vectors I inserted before?

**Comment by yhmo** [comment, marked as answer]
- Milvus has no interface to get center vector of each IVF cluster
- Center vector is not any one of the vectors you inserted. Each center vector is generated by iterate computation.
- To retrieve center vectors, one solution is add source code to export/return center vectors of IVF_FLAT index, another solution is directly read data from index file according to its format. Both solutions are not easy to implement.

**Reply by godlockin** [reply]
do we have plan to implement the related api on this feature? it's quite meaningful for us to finger out the center vector in our biz process

**Reply by yhmo** [reply]
Now for the v1.x we only fix important bugs, new feature will be ignored.
For v2.0, we collect new features and arrange our plan according to their priority. I think this feature's priority is low because it is not in main workflow.

Consider these reasons, use python to read center vectors from IVF_FLAT index is the best way. Since you don't know the file format, I can try write a python script to do this as the beginning of your work.

**Reply by godlockin** [reply]
It's quite helpful for us to have that stuff 👍

**Comment by yhmo** [comment]
The script to read center vectors of IVF_FLAT:
Set "milvus_db_path" to your milvus db path, and set the "collection_name" to your collection name, this script will find out all IVF_FLAT index files and read center vectors. 

```
import os
import struct
from milvus import Milvus, IndexType
from milvus.client.types import MetricType

SERVER_ADDR = "127.0.0.1"
SERVER_PORT = '19530'

def read_header(f):
    data = f.read(4)
    magic_num = struct.unpack("i", data)[0]
    print("magic number:", magic_num)

    data = f.read(4)
    dimension = struct.unpack("i", data)[0]
    print("dimension:", dimension)

    data = f.read(8)
    vec_count = struct.unpack("l", data)[0]
    print("vec count:", vec_count)

    # dummy
    data = f.read(8)
    data = f.read(8)

    data = f.read(1)
    is_trained = struct.unpack("?", data)[0]
    print("is_trained:", is_trained)

    # metric type, IP = 0, L2 = 1
    data = f.read(4)
    metric_type = struct.unpack("i", data)[0]
    print("metric type:", metric_type)

    return dimension

def read_centroid(filepath):
    print("========== READ INDEX DATA ==============")
    print(filepath)
    f = open(filepath, "rb")

    data = f.read(4)
    index_type = struct.unpack("i", data)[0]
    print("index type:", index_type)
    if index_type != 2:
        print("the index file is not IVF_FLAT")
        exit(1)

    data = f.read(8)
    index_name_len = struct.unpack("l", data)[0]
    print("index name length:", index_name_len)
    data = f.read(index_name_len)
    index_name = struct.unpack(str(index_name_len) + "s", data)[0]
    print("index name:", index_name)

    data = f.read(8)
    index_data_len = struct.unpack("l", data)[0]
    print("index data length:", index_data_len)

    # header of IVF_FLAT
    read_header(f)

    data = f.read(8)
    nlist = struct.unpack("l", data)[0]
    print("nlist:", nlist)

    data = f.read(8)
    nprobe = struct.unpack("l", data)[0]
    print("nprobe:", nprobe)

    # header of IVF quantizer
    dimension = read_header(f)

    # center points data size
    data = f.read(8)
    center_data_size = struct.unpack("l", data)[0]
    print("center data size:", center_data_size)

    if dimension * nlist != center_data_size:
        print("invalid index data")
        exit(1)

    vector_list = []
    for i in range(nlist):
        vector = []
        for k in range(dimension):
            data = f.read(4)
            val = struct.unpack("f", data)[0]
            vector.append(val)

        vector_list.append(vector)
        print(vector)

    print("========================================")
    return vector_list

def get_collection_stats(collection_name):
    MILVUS = Milvus(host=SERVER_ADDR, port=SERVER_PORT)
    status, stat = MILVUS.get_collection_stats(collection_name=collection_name)
    return stat

def find_index_file(segment_folder):
    dir, seg_name = os.path.split(segment_folder)
    items = os.listdir(segment_folder)
    for item in items:
        path = os.path.join(segment_folder, item)
        if os.path.isdir(path):
            continue
        else:
            dir, ext = os.path.splitext(path)
            if len(ext) != 0:
                continue

            compared = 6
            if item[:compared] == seg_name[:compared]:
                return path

    return ""

if __name__ == "__main__":
    milvus_db_path = "/tmp/milvus/db/tables/"
    collection_name = "TEST"
    stat = get_collection_stats(collection_name)
    print(stat)
    partitions = stat['partitions']
    for partition in partitions:
        segments = partition['segments']
        for segment in segments:
            if segment['index_name'] == 'IVFFLAT':
                segment_path = milvus_db_path + collection_name + "/" + segment['name']
                index_path = find_index_file(segment_path)
                if len(index_path) != 0:
                    read_centroid(index_path)
```

**Reply by godlockin** [reply]
super cool~

One question here, about the "milvus_db_path", where is it in milvus?
I can find a folder named "minio_data" in standalone deploy, minio container, but I'm not sure if that is the big shot?
how about we deploy as a cluster? the same folder exists in minio pod, but isn't it some folder take the same responsible on datanode or somewhere?

**Reply by yhmo** [reply]
This script is for v1.x
The v2.0 store files in minio, and the file format is different to v1.x.

**Reply by godlockin** [reply]
oops, so...what can I do to make the similar operations on the 2.x version? does the related api (or maybe it takes several steps) in 2.x for now? cuz the docker images were in v2.x in official sample docker files so far

**Reply by yhmo** [reply]
There is no workaround to retrieve center vectors of IVF in v2.x. I think you can create an issue to ask a new api, it is a new feature. It is a special requirement, not sure the feature can be put into plan.

**Reply by matthew-wei** [reply]
So cool,
Bug, do you use milvus 2.0?
How can i get the center vector  if i use milvus 2.0?

**Reply by yhmo** [reply]
@matthew-wei 
Milvus 2.0 storage is more complex than 1.0. Each segment is split into several parts(we call "binlog"), and wrapped by Parquet format, then be persisted into MinIO with key-value. To write a script to extract the center vectors, you need to clearly know each detail of the storage machinery.

**Reply by LLY28** [reply]
Hello, I want to find the vector corresponding to each cluster, but I haven't found the API interface yet. How to parse the index file to obtain the corresponding vector of each cluster

**Reply by yhmo** [reply]
> Hello, I want to find the vector corresponding to each cluster, but I haven't found the API interface yet. How to parse the index file to obtain the corresponding vector of each cluster

Which milvus version? If you are using 1.x, you can use my script in the comment of this discussion. If you are using 2.0, it would be difficult. Neither v1.x nor v2.0 provides such API to get center vectors.

**Reply by LLY28** [reply]
> > Hello, I want to find the vector corresponding to each cluster, but I haven't found the API interface yet. How to parse the index file to obtain the corresponding vector of each cluster
> 
> Which milvus version? If you are using 1.x, you can use my script in the comment of this discussion. If you are using 2.0, it would be difficult. Neither v1.x nor v2.0 provides such API to get center vectors.

Milvus Version is V1.0.0 . I have obtained the center vector according to your script, but I also want to obtain all the vectors corresponding to each cluster after clustering.

**Reply by matthew-wei** [reply]
> @matthew-wei Milvus 2.0 storage is more complex than 1.0. Each segment is split into several parts(we call "binlog"), and wrapped by Parquet format, then be persisted into MinIO with key-value. To write a script to extract the center vectors, you need to clearly know each detail of the storage machinery.

Thanks. Where did you get the detail of the storage machinery of Milvus V1.0 ?

**Reply by yhmo** [reply]
> > @matthew-wei Milvus 2.0 storage is more complex than 1.0. Each segment is split into several parts(we call "binlog"), and wrapped by Parquet format, then be persisted into MinIO with key-value. To write a script to extract the center vectors, you need to clearly know each detail of the storage machinery.
> 
> Thanks. Where did you get the detail of the storage machinery of Milvus V1.0 ?

Read faiss source code and milvus source code, and observe the serialize process during debugging.
